cpp/src/table/row_operators.cu

```diff
@@ -0,0 +1,207 @@
+      while (curr_col_idx > 0) {
+        auto const& prev_col = flattened[curr_col_idx - 1];
+        if (not is_nested(prev_col.type())) {
+          // We hit a column that's a leaf so seal this hierarchy
+          r_verticalized_columns.push_back(curr_col);
+          r_verticalized_col_depths.push_back(depths[curr_col_idx - 1]);
+          curr_col = prev_col;
+        } else {
+          curr_col = column_view(prev_col.type(),
+                                 prev_col.size(),
+                                 nullptr,
+                                 prev_col.null_mask(),
+                                 UNKNOWN_NULL_COUNT,
+                                 prev_col.offset(),
+                                 {curr_col});
+        }
+        --curr_col_idx;
+      }
+      verticalized_columns.insert(
+        verticalized_columns.end(), r_verticalized_columns.rbegin(), r_verticalized_columns.rend());
+      verticalized_col_depths.insert(verticalized_col_depths.end(),
+                                     r_verticalized_col_depths.rbegin(),
+                                     r_verticalized_col_depths.rend());
+      if (not column_order.empty()) {
+        new_column_order.insert(
+          new_column_order.end(), r_verticalized_columns.size(), column_order[col_idx]);
+      }
+      if (not null_precedence.empty()) {
+        new_null_precedence.insert(
+          new_null_precedence.end(), r_verticalized_columns.size(), null_precedence[col_idx]);
+      }
+    } else {
+      verticalized_columns.push_back(col);
+      verticalized_col_depths.push_back(0);
+      if (not column_order.empty()) { new_column_order.push_back(column_order[col_idx]); }
+      if (not null_precedence.empty()) { new_null_precedence.push_back(null_precedence[col_idx]); }
+    }
+  }
+  return std::make_tuple(table_view(verticalized_columns),
+                         std::move(new_column_order),
+                         std::move(new_null_precedence),
+                         std::move(verticalized_col_depths));
+}
+
+/**
+ * @brief Check a table for compatibility with lexicographic comparison
+ *
+ * Checks whether a given table contains columns of non-relationally comparable types.
+ */
+void check_lex_compatibility(table_view const& input)
+{
+  // Basically check if there's any LIST hiding anywhere in the table
+  std::function<void(column_view const&)> check_column = [&](column_view const& c) {
+    CUDF_EXPECTS(c.type().id() != type_id::LIST,
+                 "Cannot lexicographic compare a table with a LIST column");
+    if (not is_nested(c.type())) {
+      CUDF_EXPECTS(is_relationally_comparable(c.type()),
+                   "Cannot lexicographic compare a table with a column of type " +
+                     jit::get_type_name(c.type()));
+    }
+    for (auto child = c.child_begin(); child < c.child_end(); ++child) {
+      check_column(*child);
+    }
+  };
+  for (column_view const& c : input) {
+    check_column(c);
+  }
```

Comment:
Minor nit: It might be cleaner to implement `check_column` as a function returning a bool (maybe renamed to `is_lex_compatible` or so) instead of one that raises. Then you could use `std::any` (both for the main invocation and the recursive invocations inside the function) and call `CUDF_EXPECTS` if the function returns false. Not a blocker for merge, might be nice for future work though.

Reply:
I did consider it. But I wanted to have descriptive exceptions for exactly why this table is incompatible.